### PR TITLE
SymbolMapper refactor, extracted from #8435

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
This commits are extracted from #8435

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 1]
```diff
@@ -68,74 +74,98 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
         }, value);
     }
 
-    public AggregationNode map(AggregationNode node, PlanNode source)
+    public <T extends PlanNode> T map(PlanNode planNode, PlanNodeIdAllocator idAllocator, RecursionStrategy recursionStrategy)
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: * I really like the overall goal of having a general SymbolMapper facility, and using it in UnaliasSymbolReferences (or of course its iterative equivalent) as a special case.  That will also easily allow a general iterative rename-pushdown optimization.
* I don't think SymbolMapper needs to expose RecursionStrategy to its user.  Recursive uses can use generic function like this patch has, and the previous interface already worked for non-recursive uses.
* Could you explain how the SymbolMapper builder and class is a better idea than having a static function for each Node type that takes a "fromTo" map as an explicit argument?  The only thing I can come up with is that IntelliJ can prompt the user of the builder with the argument names "from" and "to", but maybe I'm missing something.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: > I don't think SymbolMapper needs to expose RecursionStrategy to its user. Recursive uses can use generic function like this patch has, and the previous interface already worked for non-recursive uses.

No problem. I can hide it Also I will restore the previous API.

> Could you explain how the SymbolMapper builder and class is a better idea than having a static function for each Node type that takes a "fromTo" map as an explicit argument? The only thing I can come up with is that IntelliJ can prompt the user of the builder with the argument names "from" and "to", but maybe I'm missing something.

`SymbolMapper` has public constructor so you do not need to use `Builder` if you do not need. To me it is clear for example in case when you already have mapping (`fromTo` map) using a `Builder` just to create a `SymbolMapper` would be at least cumbersome. However, in case where I need to do some traversal in order to collect mapping, then it is a bit easier and cleaner to collect them in such builder than collecting them in `ImmutableMap.Builder` and then pass it to `SymbolMapper`

I would prefer to avoid to put overloaded methods, one per each `PlanNode` type (times two (or three): recursive, non recursive (with singleton and list)). Because  there is a visitor used underneath and also because it is a lot of boilerplate code. The only reason I see for now is to forbid to call a `map` with a node which is not yet supported. Is this your point?

Also please take a look why this refactoring is needed.
https://github.com/prestodb/presto/pull/8435
https://github.com/prestodb/presto/pull/8435/commits/d5720446eb4ecfbda2872ae8b7886168472e509b#diff-ab20e8b431228e8711be04b33c0acbbfR168

Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: My main goal is to keep the non-recursive interface as simple as possible, even at the expense of a slightly more tedious implementation behind the interface.  I think the approach in the PR is okay, though, and I understand that you might weigh these goals differently.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: Let me try to create a better API then ;)

**Comment by kokosing** [comment]
@alandpost How about now?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 2]
```diff
@@ -626,14 +631,6 @@ else if (DeterminismEvaluator.isDeterministic(expression) && !(expression instan
             return assignments.build();
         }
 
-        private Optional<Symbol> canonicalize(Optional<Symbol> symbol)
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java]: I like what you did here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 3]
```diff
@@ -68,74 +70,108 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
         }, value);
     }
 
-    public AggregationNode map(AggregationNode node, PlanNode source)
+    public <T extends PlanNode> T map(T planNode, PlanNode source)
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: I think this is a very nice interface.  Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 4]
```diff
@@ -68,74 +70,108 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
...
-        for (Symbol symbol : outputs) {
-            Symbol canonical = map(symbol);
-            if (added.add(canonical)) {
-                builder.add(canonical);
+            ImmutableMap.Builder<Symbol, Aggregation> aggregations = ImmutableMap.builder();
+            for (Map.Entry<Symbol, Aggregation> entry : node.getAggregations().entrySet()) {
+                Symbol symbol = entry.getKey();
+                Aggregation aggregation = entry.getValue();
+
+                aggregations.put(map(symbol), new Aggregation(
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: I guess this was a pre-existing problem, but it would be great in the future if we could enhance SymbolMapper to handle the case where a node defines a new column C (such as an aggregate) which already appears in the range of the from-to map.  In that case, I think the thing to do is allocate a new symbol D, and add a mapping from C to D.  I'm thinking here about the rename-pushdown use case.  What do you think?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 5]
```diff
@@ -595,11 +596,10 @@ private void map(Symbol symbol, Symbol canonical)
             mapping.put(symbol, canonical);
         }
 
-        private Assignments canonicalize(Assignments oldAssignments)
+        private void map(Assignments assignments)
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java]: The verb "map" makes me thing it will create something and return it.  How about a name suggestive of side-effects, like "addMappings" or "gatherAliases"?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 6]
```diff
@@ -172,6 +230,25 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
             }
             return builder.build();
         }
+
+        private <T> Map<Symbol, T> mapAndDistinct(Map<Symbol, T> map)
+        {
+            Map<Symbol, T> result = new HashMap<>();
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: Could the code assert that the colliding `T`s are equal?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 7]
```diff
@@ -64,6 +64,13 @@ public Symbol map(Symbol symbol)
         return canonical;
     }
 
+    public List<Expression> map(List<Expression> values)
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: I don't see a change to SymbolMapper in the "Support ValuesNode in SymbolMapper" commit.  Is it missing code, or am I misunderstanding something?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: I think I do not understand you completely, lets go through the example.

With mapping `('a' -> 'A', 'b' - 'B')` and plan:
```
project ([a: 1, b: 3 + a],  source)
```

So in this case columns `a` and `b` exists in range of mapping, so why should I allocate new symbol and not to reuse just a mapping to produce:
```
project ([A: 1, B: 3 + A],  source)
```
This is how `SymbolMapper` behaves currently. What have I missed?

Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: I'm imagining a plan like:
```
project ([a: b, c: 3 + b],
  aggregation a = count, by b
    values (b)
```
Note that the name `'a'` gets used to label different columns in the input of the project and the output of the project.

Rename pushdown would create a mapping `('b' -> 'a')` from the project node, and then encounter trouble in the aggregation node, where the two uses of `'a'` would try to exist simultaneously in the outputs.

**Comment by kokosing** [comment]
@alandpost Can you take another look?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: currently planner should not produce such plan, as symbol allocator will encounter a conflict and so it will look like:
```
project ([a1: b, c: 3 + b],
  aggregation a = count, by b
    values (b)
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/util/PlanNodeUtils.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class PlanNodeUtils
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/util/PlanNodeUtils.java]: Callit it `ValuesNodeSymbolPrunner` to avoid creating junk drawer.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 9]
```diff
@@ -228,6 +237,20 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<Void> context)
...
+        public PlanNode visitValues(ValuesNode node, RewriteContext<Void> context)
+        {
+            ValuesNode intermediate = new ValuesNode(
+                    planNodeIdAllocationStrategy.getId(node),
+                    node.getOutputSymbols(),
+                    node.getRows().stream()
+                            .map(SymbolMapper.this::map)
+                            .collect(toImmutableList()));
+
+            Set<Symbol> added = new HashSet<>();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: explain how `Set::add` is used as predicate in comment.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java (hunk 4 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: Could the mapping function have an assertion that it hasn't run into this problem, so that if the planner changes in the future, we'll get a clear flag?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Test
+    public void testAggregationNodeSymbolMapping()
+    {
+        PlanBuilder builder = new PlanBuilder(new PlanNodeIdAllocator(), metadata);
+        ValuesNode values = builder.values(a, b);
+        AggregationNode aggregation = builder.aggregation(aggregationBuilder -> aggregationBuilder
+                .addGroupingSet(a, b)
+                .addAggregation(c, builder.expression("count(b)"), ImmutableList.of(BigintType.BIGINT))
+                .source(values));
```
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java]: Shouldn't the aggregation node also have symbols for group id, hash, and an aggregation mask?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java [hunk 11]
```diff
@@ -100,4 +109,39 @@ public void testTopNNodeSymbolMapping()
...
+
+        ProjectNode mappedProjectNode = SYMBOL_MAPPER.map(project, values);
+        assertEquals(mappedProjectNode.getOutputSymbols(), singletonList(A));
+    }
+
+    @Test
+    public void testTableScanNodeSymbolMapping()
+    {
+        PlanBuilder builder = new PlanBuilder(new PlanNodeIdAllocator(), metadata);
+        Symbol custkey = builder.symbol("custkey", BIGINT);
```
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java]: There is now an overload PlanBuilder.symbol(String) that defaults to BIGINT.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java [hunk 12]
```diff
@@ -100,4 +109,39 @@ public void testTopNNodeSymbolMapping()
         TopNNode mappedTopNNode = SYMBOL_MAPPER.map(topNNode, values);
```
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java]: Should this also check that the order symbols get mapped?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java [hunk 13]
```diff
@@ -100,4 +109,39 @@ public void testTopNNodeSymbolMapping()
...
+                        custkey, new TpchColumnHandle(custkey.getName(), BIGINT),
+                        orderkey, new TpchColumnHandle(orderkey.getName(), BIGINT)));
+
+        SymbolMapper symbolMapper = SymbolMapper.builder()
+                .put(custkey, orderkey)
+                .put(orderkey, A)
+                .build();
+
+        TableScanNode mappedTopNNode = symbolMapper.map(tableScanNode);
+        assertEquals(mappedTopNNode.getOutputSymbols(), singletonList(A));
```
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java]: Should this also check that originalConstraint gets mapped?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 14]
```diff
@@ -228,6 +237,20 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<Void> context)
                     mapAndDistinct(node.getCorrelation()));
         }
 
+        @Override
+        public PlanNode visitValues(ValuesNode node, RewriteContext<Void> context)
+        {
+            ValuesNode intermediate = new ValuesNode(
+                    planNodeIdAllocationStrategy.getId(node),
+                    node.getOutputSymbols(),
+                    node.getRows().stream()
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: Is it necessary to map the rows?  I assumed they were always literals, but I guess the types don't ensure that.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 15]
```diff
@@ -228,6 +237,20 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<Void> context)
...
+                            .map(SymbolMapper.this::map)
+                            .collect(toImmutableList()));
+
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: I think it would be good to assert that any newly colliding columns contain the same values.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java (hunk 14 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: They are... but I think it may change. This is not yet supported but correlated values subquery may refer symbols from outer queries.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java (hunk 15 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: See my answer for comment in regards to `mapAndDistinct(Map)`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: added a comment

**Comment by kokosing** [comment]
@losipiuk I had to restore `PlanNodeId` generation into `SymbolMapper`, take a look at `PushPartialAggregationThroughJoin` or `PushTopNThroughUnion`, these classes need to generate new plan node id when pushind down nodes. This is needed because they create new nodes, by mapping some another nodes.

File: presto-benchto-benchmarks/src/main/resources/sql/presto/tpch/q7.sql [hunk 16]
```diff
@@ -10,12 +10,12 @@ FROM (
          extract(YEAR FROM l.shipdate)      AS l_year,
```
Comment (alandpost) [inline presto-benchto-benchmarks/src/main/resources/sql/presto/tpch/q7.sql]: Was this accidentally included in the commit?

File: presto-benchto-benchmarks/src/main/resources/sql/presto/tpch/q7.sql [hunk 17]
```diff
@@ -10,12 +10,12 @@ FROM (
...
          l.extendedprice * (1 - l.discount) AS volume
        FROM
```
Comment (alandpost) [inline presto-benchto-benchmarks/src/main/resources/sql/presto/tpch/q7.sql]: Was this accidentally included in the commit?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+                .hashSymbol(a)
+                .groupIdSymbol(a)
+                .source(values));
+
+        AggregationNode mappedAggregation = SYMBOL_MAPPER.map(aggregation, values);
+        assertEquals(mappedAggregation.getGroupingSets(), singletonList(singletonList(A)));
+        assertEquals(mappedAggregation.getAggregations().keySet(), singleton(c));
+        assertEquals(getOnlyElement(mappedAggregation.getAggregations().values()).getCall().getArguments(), singletonList(A.toSymbolReference()));
+        assertEquals(mappedAggregation.getHashSymbol().get(), c);
+        assertEquals(mappedAggregation.getGroupIdSymbol().get(), c);
```
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java]: What about the aggregation mask?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 19]
```diff
@@ -1686,7 +1686,8 @@ else if (high != 0x4b3b4ca85a86c47aL) {
     private static byte[] reverse(final byte[] a)
     {
         final int length = a.length;
-        for (int i = length / 2; i-- != 0; ) {
+        int i = length / 2;
```
Comment (alandpost) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: This looks to me like maybe you accidentally included an unrelated change?

Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: Sorry about that once again. I have to many context switches. And sometimes things leak.

**Comment by kokosing** [comment]
@losipiuk Once you find free cycles can you please have one more look?

**Comment by kokosing** [comment]
@losipiuk Can you have another look? I would like to merge it.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+public class TestSymbolMapper
+{
+    private static final Symbol a = new Symbol("a");
+    private static final Symbol b = new Symbol("b");
+    private static final Symbol c = new Symbol("c");
+    private static final Symbol A = new Symbol("A");
+    private static final Symbol B = new Symbol("B");
+
+    private static final SymbolMapper SYMBOL_MAPPER = SymbolMapper.builder()
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java]: You could make SymbolMapper handle cycles correctly (fail). And add test for that.
Probably a followup.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 21]
```diff
@@ -68,74 +70,118 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
...
+        public Rewriter(RecursionStrategy recursionStrategy, PlanNodeIdGenerationStrategy planNodeIdGenerationStrategy)
+        {
+            this.recursionStrategy = requireNonNull(recursionStrategy, "recursionStrategy is null");
+            this.planNodeIdGenerationStrategy = planNodeIdGenerationStrategy;
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<Void> context)
+        {
+            // in case you need to have support for some plan node, please see UnaliasSymbolReferences for mapping code
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: what is that comment for?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java [hunk 22]
```diff
@@ -68,74 +70,118 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
...
     {
-        return map(node, source, node.getId());
+        return map(planNode, ImmutableList.copyOf(sources));
     }
 
-    public AggregationNode map(AggregationNode node, PlanNode source, PlanNodeIdAllocator idAllocator)
+    public <T extends PlanNode> T map(T planNode, List<PlanNode> sources)
     {
-        return map(node, source, idAllocator.getNextId());
+        return map(planNode, inject(sources), PlanNode::getId);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java]: `noRecursion` seem better name than `inject` to me

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java [hunk 23]
```diff
@@ -108,4 +116,42 @@ public void testTopNNodeSymbolMapping()
         assertEquals(mappedTopNNode.getOrderBy(), singletonList(A));
         assertEquals(mappedTopNNode.getOrderings().keySet(), singleton(A));
     }
+
+    @Test
+    public void testProjectNodeSymbolMapping()
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java]: Please add some less trivial expression than identity to the test.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/util/ValueNodesSymbolsPrunner.java [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class ValueNodesSymbolsPrunner
+{
+    private ValueNodesSymbolsPrunner()
+    {
+    }
+
+    public static PlanNode pruneValuesNode(ValuesNode node, Predicate<Symbol> predicate)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/util/ValueNodesSymbolsPrunner.java]: please add short javadoc. It is non-trivial to guess what method is doing based on name.

**Comment by kokosing** [comment]
Closing this for now as it appeared not needed for #8435.
